### PR TITLE
Refactoring compiler

### PR DESCRIPTION
**Pull request by erikvansebille**
This PR implements #869, and supports external libraries and multiple source files in the Compiler

Thanks to @CKehl, from whose #862 most of these changes were lifted

### CODE REVIEW

**Review by CKehl — APPROVED** [review]

**Review by angus-g — COMMENTED** [review]
It does seem like this will make it a bit easier to configure different compilers!

I've got a few comments about structure: there's quite a lot of duplication between different compiler configs, and they're pretty verbose.

File: parcels/compilation/codecompiler.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+            cppargs = []
+        if ldargs is None:
+            ldargs = []
+        if incdirs is None:
+            incdirs = []
+        if libdirs is None:
+            libdirs = []
+        if libs is None:
+            libs = []
+        self._compiler = "cc"
```
[the remark this answers is not in the document]
Comment (CKehl) [reply parcels/compilation/codecompiler.py:128]: with respect to the lines you marked here, I don't understand your comment. The variable `self_.compiler="cc"` is only there to initialise the variable with _some_ valid value.

File: parcels/compilation/codecompiler.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+        if cppargs is None:
+            cppargs = []
+        if ldargs is None:
+            ldargs = []
+        if incdirs is None:
+            incdirs = []
+        if libdirs is None:
+            libdirs = []
+        if libs is None:
+            libs = []
```
[the remark this answers is not in the document]
Comment (CKehl) [reply parcels/compilation/codecompiler.py:127]: could possible be done if implementing some setter-functions to the `Compiler-parameters`, yes.

File: parcels/compilation/codecompiler.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (angus-g) [reply parcels/compilation/codecompiler.py:128]: `_cppargs`, `_ldargs`, `_incdirs`, `_libdirs`, `_libs` could be refactored to a common (super?) initialiser, instead of defining them on each inheriting class

looking at it again, `_cc`, `_dynlib_ext`, `_stclib_ext`, `_obj_ext` would still need to be defined somewhere, and doesn't really make a difference whether this is on the class itself on within the initialiser

[the remark this answers is not in the document]
Comment (CKehl) [reply parcels/compilation/codecompiler.py:128]: no, so, those sets of parameters are two different shoes: `_dynlib_ext`, `_stclib_ext` and `_obj_ext` are parameters consistent for a single Operating System - hence they are implicit with the OS-specific class.

The flag parameters are specific to each compiler individually - (a) in term of e.g. if its `-Iflags` or `/Iflags` (-> Microsoft VS Compiler), and (b) thecause if defined on the user level, this compiler class then needs to 'strip' away those includes and library-links that are not on the system, or that cannot be on the system (looking for libm on windows?); (c) the cppargs need to be stripped depending on the compiler's ANSI-C compliance. Due to those difficulties, I had them still in the subclass.

Still, something slightly less verbose could be done, I agree.

File: parcels/compilation/codecompiler.py [hunk 3]
```diff
@@ -94,17 +94,14 @@ def __init__(self, cppargs=None, ldargs=None, incdirs=None, libdirs=None, libs=N
...
-            self._cppargs += Iflags
+        self._cppargs += Iflags
         self._cppargs += opt_flags + cppargs + arch_flag
         self._ldargs = ['-shared']
-        if len(Lflags) > 0:
-            self._ldargs += Lflags
-        if len(lflags) > 0:
-            self._ldargs += lflags
+        self._ldargs += Lflags
+        self._ldargs += lflags
```
[the remark this answers is not in the document]
Comment (erikvansebille) [reply parcels/compilation/codecompiler.py]: No, I haven't tried with `mpicc`, I don't have that installed. Could you try? 

And you say below that it could be made less verbose. Do you want to try that too, @CKehl?

**Review by angus-g — CHANGES_REQUESTED** [review]
This looks like it should work for the moment, although I haven't tested the extension to the non-GCC compilers. Importantly, it respects `CC` for non-MPI compilation (although in MPI mode it will *always* use `mpicc`).

Just one problem with the specification of the linker flags that I can see, otherwise seems fine!

File: parcels/particleset.py [hunk 4]
```diff
@@ -651,7 +653,7 @@ def execute(self, pyfunc=AdvectionRK4, endtime=None, runtime=None, dt=1.,
             if self.ptype.uses_jit:
                 self.kernel.remove_lib()
                 cppargs = ['-DDOUBLE_COORD_VARIABLES'] if self.lonlatdepth_dtype == np.float64 else None
-                self.kernel.compile(compiler=GNUCompiler(cppargs=cppargs))
+                self.kernel.compile(compiler=GNUCompiler(cppargs=cppargs, incdirs=[path.join(get_package_dir(), 'include'), "."]))
```
[the remark this answers is not in the document]
Comment (CKehl) [reply parcels/particleset.py:656]: It indeed requires the person writing a customized kernel interface to know what the arguments are. Given that this is an extremely rare case, as re-writing the Kernel class or subclassing it is not an everyday-casual thing, I think it's save to require it.

It is now that way because the `Compiler`-class interface can now also be used outside the code generator to JIT-compile C-code files and interlink them with the in-time code generation parts. Now, if one has pre-cooked C-source and header files and just wants to get them into `ctypes`, one doesn't require including the tools or code generator. Also, these codes are used in classes outside the kernel - such as Particles or Nodes. So, if we want to avoid compile ambiguity, we cannot compile every C-source with the parcels-include directory that the Kernel relies upon. And thus, every user of the compiler needs to explicitly state its `include` and `src` paths.

So, in short: in order to avoid strict dependencies allow allow for JIT-compiled support classes from other classes and packages, this move was inevitable.

Does it create a problem for your group ?

[the remark this answers is not in the document]
Comment (angus-g) [reply parcels/particleset.py:674]: Yeah, this is perfectly reasonable! I'm the one calling the kernel compilation directly in our case, so I can adjust for the new interface no problem.
